nrf-hal-common/src/usbd/mod.rs

```diff
@@ -0,0 +1,772 @@
+//! A `usb-device` implementation using the USBD peripheral.
+//!
+//! Difficulties:
+//! * Control EP 0 is special:
+//!   * Setup stage is put in registers, not RAM.
+//!   * Different events are used to initiate transfers.
+//!   * No notification when the status stage is ACK'd.
+
+mod errata;
+
+use crate::{
+    clocks::{Clocks, ExternalOscillator},
+    pac::USBD,
+};
+use core::sync::atomic::{compiler_fence, Ordering};
+use core::{cell::Cell, mem, ptr};
+use core::mem::MaybeUninit;
+use cortex_m::interrupt::{self, Mutex};
+use usb_device::{
+    bus::{PollResult, UsbBus, UsbBusAllocator},
+    endpoint::{EndpointAddress, EndpointType},
+    UsbDirection, UsbError,
+};
+
+fn dma_start() {
+    compiler_fence(Ordering::Release);
+}
+
+fn dma_end() {
+    compiler_fence(Ordering::Acquire);
+}
+
+struct Buffers {
+    // Ptr and size is stored separately to save space
+    in_bufs: [*mut u8; 9],
+    out_bufs: [*mut u8; 9],
+
+    // Buffers can be up to 64 Bytes since this is a Full-Speed implementation.
+    in_lens: [u8; 9],
+    out_lens: [u8; 9],
+}
+
+impl Buffers {
+    fn new() -> Self {
+        Self {
+            in_bufs: [ptr::null_mut(); 9],
+            out_bufs: [ptr::null_mut(); 9],
+            in_lens: [0; 9],
+            out_lens: [0; 9],
+        }
+    }
+}
+
+unsafe impl Sync for Buffers {}
+
+#[derive(Copy, Clone)]
+enum TransferState {
+    NoTransfer,
+    Started(u16),
+}
+
+#[derive(Copy, Clone)]
+struct EP0State {
+    direction: UsbDirection,
+    in_transfer_state: TransferState,
+    is_set_address: bool,
+}
+
+/// USB device implementation.
+pub struct Usbd<'c> {
+    periph: Mutex<USBD>,
+    // argument passed to `UsbDeviceBuilder.max_packet_size_0`
+    max_packet_size_0: u16,
+    unalloc_buffers: &'static mut [u8],
+    bufs: Buffers,
+    used_in: u8,
+    used_out: u8,
+    iso_in_used: bool,
+    iso_out_used: bool,
+    ep0_state: Mutex<Cell<EP0State>>,
```

Comment:
Heh, I have literally the same field name and struct name in my non-working copy of this :-)
I only got as far as effectively the same thing as your `TransferState`, but structured a little differently.